easypy/concurrency.py

```diff
@@ -1268,3 +1281,165 @@ def __call__(cls, *args, **kwargs):
     @synchronized
     def get_instance(cls):
         return cls._instances.get(cls)
+
+
+class SynchronizationCoordinatorWrongWait(TException):
+    template = "Task is waiting on {this_file}:{this_line} instead of {others_file}:{others_line}"
+
+
+class SynchronizationCoordinator(object):
+    """
+    Synchronization helper for functions that run concurrently.
+
+        sync = SynchronizationCoordinator(5)
+
+        def foo(a):
+            sync.wait_for_everyone()
+            sync.collect_and_call_once(a, lambda a_values: print(a))
+
+        MultiObject(range(5)).call(foo)
+
+    When MultiObject/concurrent_map/sync runs a function with a _sync=SYNC
+    argument, it will replace it with a proper SynchronizationCoordinator instance:
+
+        def foo(a, _sync=SYNC):
+            _sync.wait_for_everyone()
+            _sync.collect_and_call_once(a, lambda a_values: print(a))
+
+        MultiObject(range(5)).call(foo)
+    """
+
+    def __init__(self, num_participators):
+        self.num_participators = num_participators
+        self._reset_barrier()
+        self._lock = threading.Lock()
+        self._call_once_collected_param = []
+        self._call_once_function = None
+        self._call_once_result = None
+
+        self._wait_context = None
+
+    def _reset_barrier(self):
+        self.barrier = threading.Barrier(self.num_participators, action=self._post_barrier_action)
+
+    def _post_barrier_action(self):
+        if self.num_participators != self.barrier.parties:
+            self._reset_barrier()
+        self._wait_context = None
+
+        if self._call_once_function:
+            call_once_function = self._call_once_function
+            self._call_once_function = None
+            collected_param = self._call_once_collected_param
+            self._call_once_collected_param = []
+
+            try:
+                self._call_once_result = call_once_function(collected_param)
+            except BaseException as e:
+                self._reset_barrier()
+                self._call_once_result = e
+                raise
+
+    def wait_for_everyone(self):
```

Comment:
There should be a `timeout` here I think, in case some rogue thread takes too long.
An TimeoutException raised here would rise out of all of the waiting threads up to `async`, which should somehow know to abandon the lagging thread (and allow it to crash (?) on the broken barrier if it ever reaches it finally)

Reply:
Adding a timeout is easy - making it abandon the threads that lag behind will be harder (unless I split the coordinator, or enforce usage of thread identities). I'd rather just invalidate the entire coordinator.